pkg/controller/management/secrets/secrets.go

```diff
@@ -0,0 +1,221 @@
+	client client.Client
+}
+
+// SetupReconcilerWithManager initializes a reconciler for Secret resources
+// and registers it with the provided Manager.
+func SetupReconcilerWithManager(
+	ctx context.Context,
+	kargoMgr manager.Manager,
+	cfg ReconcilerConfig,
+) error {
+	err := ctrl.NewControllerManagedBy(kargoMgr).
+		For(&corev1.Secret{}).
+		WithEventFilter(
+			predicate.Funcs{
+				DeleteFunc: func(event.DeleteEvent) bool {
+					// We're not interested in any deletes
+					return false
+				},
+			},
+		).
+		Complete(newReconciler(kargoMgr.GetClient(), cfg))
+	if err == nil {
+		logging.LoggerFromContext(ctx).Info(
+			"Initialized Secrets reconciler",
+			"sourceNamespace", cfg.SourceNamespace,
+			"destinationNamespace", cfg.DestinationNamespace,
+		)
+	}
+	return err
+}
+
+func newReconciler(kubeClient client.Client, cfg ReconcilerConfig) *reconciler {
+	return &reconciler{
+		cfg:    cfg,
+		client: kubeClient,
+	}
+}
+
+// Reconcile is part of the main Kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *reconciler) Reconcile(
+	ctx context.Context,
+	req ctrl.Request,
+) (ctrl.Result, error) {
+	logger := logging.LoggerFromContext(ctx).WithValues(
+		"secret.name", req.Name,
+		"secret.namespace", req.Namespace,
+	)
+
+	// Validate that the Secret is from the source namespace. If it's from a
+	// different namespace, we should not reconcile it.
+	if req.Namespace != r.cfg.SourceNamespace {
+		logger.Debug(
+			"ignoring secret from unexpected namespace",
+			"expectedNamespace", r.cfg.SourceNamespace,
+			"actualNamespace", req.Namespace,
+		)
+		return ctrl.Result{}, nil
+	}
+
+	logger.Debug("reconciling Secret")
+
+	// Get the Secret from the source namespace
+	srcSecret := &corev1.Secret{}
+	if err := r.client.Get(
+		ctx,
+		types.NamespacedName{
+			Namespace: r.cfg.SourceNamespace,
+			Name:      req.Name,
+		},
+		srcSecret,
+	); err != nil {
+		// Ignore if not found. This can happen if the Secret was deleted after the
+		// current reconciliation request was issued.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if !srcSecret.DeletionTimestamp.IsZero() {
+		logger.Debug("source Secret is being deleted; handling deletion")
+		return ctrl.Result{}, r.handleDelete(ctx, srcSecret)
+	}
+
+	// Ensure the Secret has a finalizer and requeue if it was added.
+	// The reason to requeue is to ensure that a possible deletion of the Secret
+	// directly after the finalizer was added is handled without delay.
+	if ok, err := api.EnsureFinalizer(ctx, r.client, srcSecret); ok || err != nil {
+		logger.Debug("ensured finalizer on source Secret; requeuing")
+		return ctrl.Result{RequeueAfter: 100 * time.Millisecond}, err
+	}
+
+	destSecret := srcSecret.DeepCopy()
+
+	// Remove the finalizer that was just copied over so that deletes of new
+	// secrets won't be blocked by anything.
+	controllerutil.RemoveFinalizer(destSecret, kargoapi.FinalizerName)
```

Comment:
`controllerutil.RemoveFinalizer()` returns an `error` that you're not capturing. If you were, I suspect you'd find this call is actually returning a non-nil error.

`controllerutil.RemoveFinalizer()` seems to go as far as patching the resource. If this is a first-time sync of the source resource, this would fail with a not-found error.

Rather than starting to capture the error, I think we don't need to use `controllerutil.RemoveFinalizer()` and can get away with simply removing our finalizer if it's found in the list of finalizers.